BABEL-1199 Support DATABASE_PRINCIPAL_ID() T-SQL function

pratikzode opened the pull request:
SELECT DATABASE_PRINCIPAL_ID();
Retrieve the ID of the current user.

SELECT DATABASE_PRINCIPAL_ID(‘principal_name’);
Retrieve the ID of the specific database principal.

Modified the file so that when called with NULL as argument it will return NULL and also modified for trailing spaces that will give proper output instead of NULL.

Previous PR: https://code.amazon.com/packages/RDSManfredBabelExtensions/commits/bd1f0bef054919c2a27498c55407552f9f1f9830

Task: BABEL-1199
Signed-off-by: Pratik Zode [pzode@amazon.com](mailto:pzode@amazon.com)

pratikzode committed:
BABEL-1199 Support DATABASE_PRINCIPAL_ID() T-SQL function

SELECT DATABASE_PRINCIPAL_ID();
Retrieve the ID of the current user.

SELECT DATABASE_PRINCIPAL_ID(‘principal_name’);
Retrieve the ID of the specific database principal.

Modified the file so that when called with NULL as argument it will return NULL and also modified for trailing spaces that will give proper output instead of NULL.

Task: BABEL-1199
Signed-off-by: pratikzode <pratik.zode@gmail.com>

macwoody commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -1739,26 +1739,15 @@ CREATE OR REPLACE FUNCTION sys.servicename()
+RETURNS OID
+AS 'babelfishpg_tsql', 'user_id_noarg'
+LANGUAGE C IMMUTABLE PARALLEL SAFE;
 
```
plz change user_id function as well. user_id should show same result as this function

xhfanhe commented on test/JDBC/expected/BABEL-1199-vu-verify.out:
```diff
@@ -102,3 +102,27 @@ int
 <NULL>
```
This shouldn't return a NULL output, please fix this one.

macwoody commented on contrib/babelfishpg_tsql/src/rolecmds.c:
```diff
@@ -712,42 +712,69 @@ PG_FUNCTION_INFO_V1(user_id);
-	HeapTuple	auth_tuple;
-	Form_pg_authid authform;
-	Oid			ret;
+    char    *user_input;
```
plz use tab instead of spaces

lejaokri commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.1.0--3.2.0.sql:
```diff
@@ -441,26 +441,15 @@ ALTER FUNCTION sys.json_modify RENAME TO json_modify_deprecated_in_3_2_0;
 CALL sys.babelfish_drop_deprecated_object('function', 'sys', 'json_modify_deprecated_in_3_2_0');
 
-CREATE OR REPLACE FUNCTION sys.database_principal_id(IN user_name sys.sysname DEFAULT NULL)
+CREATE OR REPLACE FUNCTION sys.database_principal_id(IN user_name sys.sysname)
```
think you need to rename instead? see https://jira.rds.a2z.com/browse/BABEL-3260 for more details

I think the problem is when a customer is using `sys.database_principal_id` without any parameters and once they update they wont be able to call the function without any parameter anymore (could even fail upgrade/update if I recall correctly)

xhfanhe replied:
Please move the code block several lines up so that we don't interrupt json_modify

macwoody commented on contrib/babelfishpg_tsql/src/rolecmds.c:
```diff
@@ -712,42 +712,69 @@ PG_FUNCTION_INFO_V1(user_id);
 
-	if (!user_input)
-		PG_RETURN_OID(GetUserId());
+    if (!user_input)
```
user_input could not be NULL since you defined the function as strict. 
Even if it can, expected result is NULL, not current user id .

macwoody commented on contrib/babelfishpg_tsql/src/rolecmds.c:
```diff
@@ -712,42 +712,69 @@ PG_FUNCTION_INFO_V1(user_id);
+    auth_tuple = SearchSysCache1(AUTHNAME, CStringGetDatum(user_name));
 
-	ReleaseSysCache(auth_tuple);
+    if (!HeapTupleIsValid(auth_tuple))
```
release sys cache before returning

macwoody commented on contrib/babelfishpg_tsql/src/rolecmds.c:
```diff
@@ -712,42 +712,69 @@ PG_FUNCTION_INFO_V1(user_id);
+    Form_pg_authid authform;
+    Oid ret;
+
+    auth_tuple = SearchSysCache1(AUTHNAME, CStringGetDatum(GetUserNameFromId(GetUserId(), false)));
```
GetUserId() = authform->oid in your case.  
No need to search again, just return it. 

refer to line 730.

pratikzode committed:
BABEL-1199 Modify test case,user_id_noarg

Task: BABEL-1199
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
and

pratikzode committed:
BABEL-1199 Modify babelfish_tsql-3.1–3.2

Task: BABEL-1199
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
BABEL-1199 Modify existing functions

Have modified user_id and did an alignment modification in rolecmds.c file

Task: BABEL-119
9Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
BABEL-1199 Modify rolecmds.c file

Task: BABEL-1199
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
BABEL-1199 Alignment in rolecmds.c file

Task: BABEL-1199
Signed-off-by: pratikzode <pratik.zode@gmail.com>

xhfanhe commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.1.0--3.2.0.sql:
```diff
@@ -108,6 +108,30 @@ CREATE AGGREGATE sys.VARP(float8) (
+AS 'babelfishpg_tsql', 'user_id_noarg'
+LANGUAGE C IMMUTABLE PARALLEL SAFE;
+
+ALTER FUNCTION sys.user_id(IN user_name TEXT DEFAULT NULL) RENAME TO babelfish_sys_user_id_deprecated_in_3_2_0;
```
DEFAULT is not accepted in PG ALTER FUNCTION syntax. See https://www.postgresql.org/docs/current/sql-alterfunction.html. And we should just add the deprecated string after the function name to avoid confusion.
`ALTER FUNCTION sys.user_id(TEXT) RENAME TO user_id_deprecated_in_3_2_0;`

xhfanhe commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.1.0--3.2.0.sql:
```diff
@@ -108,6 +108,30 @@ CREATE AGGREGATE sys.VARP(float8) (
+AS 'babelfishpg_tsql', 'user_id_noarg'
+LANGUAGE C IMMUTABLE PARALLEL SAFE;
+
+CALL sys.babelfish_drop_deprecated_object('function', 'sys', 'babelfish_sys_user_id_deprecated_in_3_2_0');
```
`CALL sys.babelfish_drop_deprecated_object('function', 'sys', 'user_id_deprecated_in_3_2_0');`

pratikzode committed:
BABEL-1199 Modify babelfish_tsql-3.1–3.2

Modified the alter function

Task: BABEL-1199
Signed-off-by: pratikzode <pratik.zode@gmail.com>

xhfanhe commented:
Test is failing at expected_dependency. Please update expected_dependency.out accordingly

pratikzode committed:
BABEL-1199 Modified expected_dependency file

Deleted Function sys.user_id(text)and added Function sys.user_id() Function sys.user_id(sys.sysname)

Task: BABEL-1199
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
BABEL-1199 Modified expected_dependency file

Task: BABEL-1199
Signed-off-by: pratikzode <pratik.zode@gmail.com>

surendravishnoi commented on test/python/expected/upgrade_validation/expected_dependency.out:
```diff
@@ -668,7 +668,8 @@ Function sys.uniqueidentifier_cmp(sys.uniqueidentifier,sys.uniqueidentifier)
 Function sys.uniqueidentifier_sqlvariant(sys.uniqueidentifier)
 Function sys.update(text)
-Function sys.user_id(text)
+Function sys.user_id()
```
Please add upgrade tests to verify views/functions/procs dependency on this function work as expected. This is important as we doing a rename here.

xhfanhe commented on test/JDBC/input/BABEL-1199-vu-verify.sql:
```diff
@@ -35,5 +35,11 @@ GO
+SELECT DATABASE_PRINCIPAL_ID('test_me')
+GO
+
+SELECT DATABASE_PRINCIPAL_ID(NULL)
```
Please add dependency tests for the new user_id function as Surendra suggested. And I'll suggest you put all these tests in JDBC/input/functions directory, and name them after the functions, so that others can easily understand the test context from the title.

pratikzode committed:
BABEL-1199 Modify Test Cases

Have added dependency tests for the new user_id function.
Have moved the test directory to JDBC/input/functions.
Have modified the latest upgrade script.
Deleted the previous files BABEL-1199 and rename sys.userid

Task: BABEL-1199
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
BABEL-1199 Modify Test Cases

Have modified the latest upgrade script from sys.userid to sys-userid

Task: BABEL-1199
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
BABEL-1199 Modify Test Cases

Added userid test cases

Task: BABEL-1199
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
BABEL-1199 Modify Test Cases

Added userid test cases

Task: BABEL-1199
Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
BABEL-1199 Modify expected_depencency file

Remove Function sys.user_id(sys.sysname) from expected_depencency

Task: BABEL-1199
Signed-off-by: pratikzode <pratik.zode@gmail.com>
